### PR TITLE
Merge common components and services to 2.x

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
* feat: split to common change

Signed-off-by: suzhou <suzhou@amazon.com>

* feat: update

Signed-off-by: suzhou <suzhou@amazon.com>

Signed-off-by: suzhou <suzhou@amazon.com>

### Description
[Describe what this change achieves]

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/components/JSONDiffEditor/JSONDiffEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            // do nothing
+          }
+        }}
+        title={`editor-is-ready-${isReady}`}
+        data-test-subj={`${others["data-test-subj"] || "jsonEditor"}-valueDisplay`}
+      />
+      <div style={{ display: "flex", marginBottom: 12 }}>
+        <div style={{ flexGrow: 1 }}>
+          <CustomFormRow label="Original" helpText="The original value">
+            <></>
```
Comment (amsiglan) [inline public/components/JSONDiffEditor/JSONDiffEditor.tsx:107]: extra empty fragment?

File: public/components/JSONDiffEditor/JSONDiffEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+          </CustomFormRow>
+        </div>
+        <div style={{ flexGrow: 1 }}>
+          <CustomFormRow label="Modified" helpText="The value you modified">
+            <></>
```
Comment (amsiglan) [inline public/components/JSONDiffEditor/JSONDiffEditor.tsx:112]: extra empty fragment

File: public/components/JSONDiffEditor/JSONDiffEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        }
+      }),
+    getValue: () => valueRef.current,
+    setValue: (val: string) => setAllValue(val),
+  }));
+
+  return (
+    <>
+      <textarea
+        style={{ display: "none" }}
```
Comment (amsiglan) [inline public/components/JSONDiffEditor/JSONDiffEditor.tsx:91]: What is this invisible text area used for?

File: public/components/JSONDiffEditor/JSONDiffEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    setAllValue(value);
+  }, [value, setAllValue]);
+
+  useEffect(() => {
+    document.body.addEventListener("click", onClickOutsideHandler.current);
+    editorRef.current?.getDomNode().addEventListener("click", onClickContainer.current);
+    editorRef.current?.getModifiedEditor().getDomNode()?.setAttribute("data-test-subj", "codeEditorContainer");
+    return () => {
+      document.body.removeEventListener("click", onClickOutsideHandler.current);
+      editorRef.current?.getDomNode().addEventListener("click", onClickContainer.current);
```
Comment (amsiglan) [inline public/components/JSONDiffEditor/JSONDiffEditor.tsx:69]: Should this be `removeEventListener`?

File: public/components/JSONDiffEditor/JSONTextArea.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  onChange?: (value: JSONDiffEditorProps["value"]) => void;
+  "data-test-subj"?: string;
+  disabled?: boolean;
+}
+
+const JSONDiffEditor = forwardRef(({ value, onChange, ...others }: JSONDiffEditorProps, ref: React.Ref<IJSONEditorRef>) => {
+  const [confirmModalVisible, setConfirmModalVisible] = useState(false);
+  const [editorValue, setEditorValue] = useState(value);
+  const valueRef = useRef(editorValue);
+  valueRef.current = editorValue;
```
Comment (amsiglan) [inline public/components/JSONDiffEditor/JSONTextArea.tsx:22]: why do we need this line if we already set the initial value in above line?

File: public/components/JSONDiffEditor/JSONTextArea.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        title={`editor-is-ready-true`}
+        data-test-subj={`${others["data-test-subj"] || "jsonEditor"}-valueDisplay`}
+      />
+      {confirmModalVisible && (
+        <EuiFormRow
+          fullWidth
+          isInvalid={confirmModalVisible}
+          error="Your input does not match the validation of json format, please fix the error line with error aside."
+        >
+          <></>
```
Comment (amsiglan) [inline public/components/JSONDiffEditor/JSONTextArea.tsx:69]: use for empty fragment?

**Comment by amsiglan** [comment]
Please break down this PR into smaller PRs. It is difficult to review these many files

File: public/JobHandler/index.tsx [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+export const destroyListener = (eventName: string, callback: () => void) => {
+  window.removeEventListener(eventName, callback);
+};
+
+export function JobHandlerRegister(core: CoreSetup) {
+  const indexService = new IndexService(core.http);
+  const commonService = new CommonService(core.http);
+  jobSchedulerInstance.addCallback({
+    callbackName: "callbackForReindex",
+    callback: async (job: ReindexJobMetaData) => {
```
Comment (amsiglan) [inline public/JobHandler/index.tsx:57]: extract this callback into a separate method/function.

File: public/components/BetterComboBox/index.tsx [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { EuiComboBox, EuiComboBoxProps } from "@elastic/eui";
+
+export default class BetterComboBox<T> extends EuiComboBox<T> {
```
Comment (amsiglan) [inline public/components/BetterComboBox/index.tsx:3]: The name `BetterComboBox` doesn't describe what enhacement is made, can we use a name that indicates the function of this extension?

Comment (amsiglan) [reply public/components/BetterComboBox/index.tsx:3]: similarly for `BetterTooltip`

**Review by amsiglan — APPROVED** [review]

File: public/components/JSONDiffEditor/JSONDiffEditor.tsx (hunk 1 above)
Comment (SuZhou-Joe) [reply public/components/JSONDiffEditor/JSONDiffEditor.tsx:107]: The EuiFormRow requires props.children so we have to add an empty fragment to pass the type check.

File: public/components/JSONDiffEditor/JSONDiffEditor.tsx (hunk 3 above)
Comment (SuZhou-Joe) [reply public/components/JSONDiffEditor/JSONDiffEditor.tsx:91]: It's used for check the editor value and set the value when we run unit tests.

File: public/components/JSONDiffEditor/JSONDiffEditor.tsx (hunk 2 above)
Comment (SuZhou-Joe) [reply public/components/JSONDiffEditor/JSONDiffEditor.tsx:112]: The EuiFormRow requires props.children so we have to add an empty fragment to pass the type check.

File: public/components/JSONDiffEditor/JSONTextArea.tsx (hunk 5 above)
Comment (SuZhou-Joe) [reply public/components/JSONDiffEditor/JSONTextArea.tsx:22]: There is a closure problem for react hooks, you can see there is a `getValue` method which gives the latest value of the editor, if we do not set it to a ref, when we call ref.getValue the result will always be the initial value.

File: public/components/JSONDiffEditor/JSONTextArea.tsx (hunk 6 above)
Comment (SuZhou-Joe) [reply public/components/JSONDiffEditor/JSONTextArea.tsx:69]: The EuiFormRow requires props.children so we have to add an empty fragment to pass the type check.

File: public/JobHandler/index.tsx (hunk 7 above)
Comment (SuZhou-Joe) [reply public/JobHandler/index.tsx:57]: > 



> Please break down this PR into smaller PRs. It is difficult to review these many files

Sorry, but it is a smaller PR already(We split a huge PR into common/index-management/...) and the common one is hard to split again. We won't make such a big PR next time.

File: public/components/JSONDiffEditor/JSONDiffEditor.tsx (hunk 4 above)
Comment (SuZhou-Joe) [reply public/components/JSONDiffEditor/JSONDiffEditor.tsx:69]: Fixed on

File: public/components/BetterComboBox/index.tsx (hunk 8 above)
Comment (SuZhou-Joe) [reply public/components/BetterComboBox/index.tsx:3]: Fixed on https://github.com/opensearch-project/index-management-dashboards-plugin/commit/70dd00514c00fd9789bef132ebf981ef9eda0ffd

File: public/JobHandler/index.tsx (hunk 7 above)
Comment (SuZhou-Joe) [reply public/JobHandler/index.tsx:57]: break down callbacks to seperate files.
https://github.com/opensearch-project/index-management-dashboards-plugin/commit/e7184e229f17eb720451d569afdf436f245aca04

**Pull request closed without merging**